criu/files-reg.c

```diff
@@ -1380,12 +1387,472 @@ static bool should_check_size(int flags)
+		*iter = 0;
+}
+
+static inline void calculate_checksum_iterator_next(int *iter, int *low_bound,
+							int *up_bound, size_t file_size,
+							unsigned char **file_header,
+							const int fd)
+{
+	if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_EVERY) {
+		*iter += 1;
+	} else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FIRST) {
+		*iter += 1;
+	} else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_PERIOD) {
+		*iter += opts.file_validation_chksm_parameter;
+
+		/*
+		 * If the new iterator position is outside the mapped region (Bytes in the range 
+		 * low_bound and up_bound) of the file, the current region needs to be unmapped
+		 * and the next portion (At most the next 10MB) of the file needs to be mapped.
+		 */
+		if (*iter >= *up_bound && *up_bound != file_size) {
+			munmap(*file_header, *up_bound - *low_bound);
+			*iter -= *up_bound;
+			*low_bound = *up_bound;
+			*up_bound = min_t(size_t, file_size, *up_bound + CHKSM_CHUNK_SIZE);
+			*file_header = (unsigned char *) mmap(0, *up_bound - *low_bound,
+						PROT_READ, MAP_PRIVATE | MAP_FILE, fd, *low_bound);
+		}
+	}
+}
+
+static inline bool calculate_checksum_iterator_stop_condition(int iter)
+{
+	if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_EVERY) {
+		return false;
+	} else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_FIRST) {
+		if (iter >= opts.file_validation_chksm_parameter)
+			return true;
+		else
+			return false;
+	} else if (opts.file_validation_chksm_config == FILE_VALIDATION_CHKSM_PERIOD) {
+		return false;
+	}
+
+	return true;
+}
+
+/*
+ * Does the actual work of calculating the CRC32C checksum of "some"
+ * parts of the file. These "some" parts depend on the configuration chosen.
+ * By default, the first 1024 bytes of the file or the entire file, which ever
+ * is smaller, is considered.
+ * Returns true if the checksum could be obtained, else it returns false.
+ */
+static bool calculate_checksum(const int fd, const struct stat *fd_status,
+				u32 *checksum)
+{
+	int bit_count, up_bound, low_bound = 0, i = 0, num_iter = 0;
+	u32 byte, mask;
+	unsigned char *file_header;
+
+	/* At most, the first 10MB of the file is mapped. */
+	up_bound = min_t(size_t, fd_status->st_size, CHKSM_CHUNK_SIZE);
+	file_header = (unsigned char *) mmap(0, up_bound,
+						PROT_READ, MAP_PRIVATE | MAP_FILE, fd, 0);
+	if (file_header == MAP_FAILED) {
+		pr_warn("Couldn't mmap file with fd %d", fd);
+		return false;
+	}
+
+	*checksum = 0xFFFFFFFF;
+	calculate_checksum_iterator_init(&i);
+
+	/*
+	 * The number of iterations is limited to 10485760 times for performance reasons.
+	 * 10485760 was chosen because regardless of the checksum configuration chosen,
+	 * at most 10MB worth of data of the file is collectively considered in the
+	 * calculation of the checksum since the checksum parameter can also never exceed
+	 * 10485760.
```

Comment:
10485760 has to be configurable

Reply:
I removed the limit for the number of the iterations. Now the problem is what if the user chooses to checksum the entire file and the files turn out to be huge? It would take a very long time to calculate the checksum using all the bytes.